src/BlockTypes/OrderConfirmation/AbstractOrderConfirmationBlock.php

```diff
@@ -0,0 +1,207 @@
+	 *
+	 * @return string | void Rendered block output.
+	 */
+	protected function render( $attributes, $content, $block ) {
+		if ( ! empty( $attributes['isPreview'] ) ) {
+			$order      = $this->get_preview_order();
+			$permission = 'full';
+		} else {
+			$order      = $this->get_order();
+			$permission = $this->get_view_order_permissions( $order );
+		}
+
+		$block_content      = $order ? $this->render_content( $order, $permission, $attributes ) : $this->render_content_fallback();
+		$classname          = $attributes['className'] ?? '';
+		$classes_and_styles = StyleAttributesUtils::get_classes_and_styles_by_attributes( $attributes );
+
+		if ( isset( $attributes['align'] ) ) {
+			$classname .= " align{$attributes['align']}";
+		}
+
+		return sprintf(
+			'<div class="wc-block-%5$s %1$s %2$s" style="%3$s">%4$s</div>',
+			esc_attr( $classes_and_styles['classes'] ),
+			esc_attr( $classname ),
+			esc_attr( $classes_and_styles['styles'] ),
+			$block_content,
+			esc_attr( $this->block_name )
+		);
+	}
+
+	/**
+	 * This renders the content of the block within the wrapper. The permission determines what data can be shown under
+	 * the given context.
+	 *
+	 * @param \WC_Order $order Order object.
+	 * @param string    $permission Permission level for viewing order details.
+	 * @param array     $attributes Block attributes.
+	 * @return string
+	 */
+	abstract protected function render_content( $order, $permission = false, $attributes = [] );
+
+	/**
+	 * This is what gets rendered when the order does not exist. Renders nothing by default, but can be overridden by
+	 * child classes.
+	 *
+	 * @return string
+	 */
+	protected function render_content_fallback() {
+		return '';
+	}
+
+	/**
+	 * Get current order.
+	 *
+	 * @todo Check auth.
+	 *
+	 * @return \WC_Order|null
+	 */
+	protected function get_order() {
+		$order_id = absint( get_query_var( 'order-received' ) );
+
+		if ( $order_id ) {
+			return wc_get_order( $order_id );
+		}
+
+		return null;
+	}
+
+	/**
+	 * View mode for order details based on the order, current user, and settings.
+	 *
+	 * Possible values are:
+	 * - "full" user can view all order details.
+	 * - "limited" user can view some order details, but no PII. This may happen for example, if the user checked out as a guest.
+	 * - false user cannot view order details.
+	 *
+	 * @param \WC_Order|null $order Order object.
+	 * @return "full"|"limited"|false
+	 */
+	protected function get_view_order_permissions( $order ) {
+		if ( ! $order ) {
+			return false;
+		}
+
+		if ( is_user_logged_in() ) {
+			// If logged in, check the user owns the order.
+			return $this->is_current_customer_order( $order ) ? 'full' : false;
+		}
+
+		// If the user is logged out, check the order key for validity.
+		if ( $this->allow_guest_checkout() && $this->is_current_customer_order( $order ) && $this->has_valid_order_key( $order ) ) {
+			return $this->order_matches_session( $order ) ? 'full' : 'limited';
+		}
```

Comment:
What are your thoughts on placing the comment outside the if-statement so that it's consistent with the comment of the next if-statement?

```suggestion
		// If logged in, check the user owns the order.
		if ( is_user_logged_in() ) {
			return $this->is_current_customer_order( $order ) ? 'full' : false;
		}

		// If the user is logged out, check the order key for validity.
		if ( $this->allow_guest_checkout() && $this->is_current_customer_order( $order ) && $this->has_valid_order_key( $order ) ) {
			return $this->order_matches_session( $order ) ? 'full' : 'limited';
		}
```